Drop `versionize` support, substitute with `serde`

bchalios opened the pull request:
## Changes

Remove support for `Versionize` from Firecracker and substitute it with `serde`. We use `bincode` for the actual (de)serialization. Also, define a snapshot format version that is decoupled from Firecracker version. The newly defined version of the snapshot format will follow semantic versioning and it will start from version `1.0.0`.

This PR breaks compatibility with all previous snapshot versions. From now on, snapshot compatibility will be reflected to the snapshot version. Changes in the major release number are backwards incompatible, changes in the minor release number are backwards compatible and changes in patch releases are always compatible.

That said, we are currently using `bincode` for the serialization, which means that all changes in the snapshot format will be backwards incompatible (they will imply a major version number bump). However, if the need arises we will investigate alternatives to `bincode` that would allow greater flexibility for backwards compatibility.

This change makes a lot of our integration tests redundant.

## Reason

We used `Versionize` to allow for extended backwards compatibility for Firecracker snapshots. In practice we did not find this feature useful, so we paid the code maintenance and testing complexity cost without a measurable benefit.

## License Acceptance

By submitting this pull request, I confirm that my contribution is made under
the terms of the Apache 2.0 license. For more information on following
Developer Certificate of Origin and signing off your commits, please check
[`CONTRIBUTING.md`][3].

## PR Checklist

- [x] If a specific issue led to this PR, this PR closes the issue.
- [x] The description of changes is clear and encompassing.
- [x] Any required documentation changes (code and docs) are included in this PR.
- [x] API changes follow the [Runbook for Firecracker API changes][2].
- [x] User-facing changes are mentioned in `CHANGELOG.md`.
- [x] All added/changed functionality is tested.
- [x] New `TODO`s link to an issue.
- [x] Commits meet [contribution quality standards](https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md#contribution-quality-standards).

---

- [ ] This functionality cannot be added in [`rust-vmm`][1].

[1]: https://github.com/rust-vmm
[2]: https://github.com/firecracker-microvm/firecracker/blob/main/docs/api-change-runbook.md
[3]: https://github.com/firecracker-microvm/firecracker/blob/main/CONTRIBUTING.md

bchalios commented:
This PR depends on the equivalent PR on kvm-bindings side: https://github.com/firecracker-microvm/kvm-bindings/pull/21

bchalios commented:
I deliberately split up the commits that switch types from `Versionize` to `serde` so that I facilitate the review. but I am happy to merge these in a single commit if we decide its cleaner that way.

roypat commented on docs/snapshotting/versioning.md:
```diff
@@ -61,59 +35,37 @@ emulation, KVM and vCPUs) with 2 exceptions - serial emulation and vsock backend
+format will cause `MAJOR` version number bump, whereas backwards compatible
+changes will cause a `MINOR` version number bump.
+
+For example, that means that a Firecracker binary that supports snapshot data
+format `x.y.z` will be able to load all snapshot files with version `x.Y.z`
+where `Y <= y`.
```
Wait, is this really semantic versioning? Shouldn't a Firecracker binary that declares support for snapshot version format, say, 2.1.0 also be able to load version, say, 2.5.0?

Also, this should say something like "will be able to load all snapshot files with version `x.Y.Z` with `Y<y` or `Y=y and Z<=z`.

bchalios replied:
I removed mentions of semantic versioning. In any case, with bincode we will not be able to make any backwards compatible changes to the snapshot format. If in the future we do, we should define this properly, for the moment we just mention the version format `MAJOR.MINOR.PATCH` which is generic enough to describe whatever backwards compatibility we need.

For the record, SchemaVer (which is based on semantic versioning) seems to fit well to what we have in mind: https://snowplow.io/blog/introducing-schemaver-for-semantic-versioning-of-schemas/

roypat replied:
Sounds good!

ShadowCurse commented on src/vmm/src/arch/aarch64/regs.rs:
```diff
@@ -473,7 +425,7 @@ impl<'a> Aarch64RegisterRefMut<'a> {
 /// else.
-#[derive(Debug, Clone, Copy, Versionize)]
+#[derive(Debug, Clone, Copy)]
 pub struct Aarch64RegisterOld {
```
It is used to restore old snap versions. Considering we are removing versionize, we can remove this type + remove `old_regs` field from `VcpuState` for aarch64.

bchalios replied:
done

bchalios commented on src/vmm/src/arch/aarch64/gic/regs.rs:
```diff
@@ -80,7 +52,7 @@ pub(crate) trait MmioReg {
 pub(crate) trait VgicRegEngine {
     type Reg: MmioReg;
-    type RegChunk: Clone + Default + Versionize;
+    type RegChunk: Clone + Default + Serialize + for<'a> Deserialize<'a>;
```
silly question probably, but is it passing on `aarch64`?

roypat replied:
Yeah, `cargo check --aarch64-unknown-linux-gnu` passes. We also dont implement serialize and deserialize for any of the types that implement the trait. I think it works because we dont ever use this trait in a generic fashion during (de)serialization, and all concrete types for which it is implemented provide a `RegChunk` type that is serializable.

bchalios replied:
done!

bchalios commented on src/vmm/src/arch/aarch64/regs.rs:
```diff
@@ -276,53 +275,6 @@ impl Aarch64RegisterVec {
-    {
-        let inner = Aarch64RegisterVecInner::deserialize(reader, version_map, source_version)?;
-        let mut total_size: usize = 0;
-        for id in inner.ids.iter() {
-            let reg_size = reg_size(*id);
-            if RegSize::U2048_SIZE < reg_size {
-                return Err(VersionizeError::Deserialize(
-                    "Failed to deserialize aarch64 registers. Registers bigger then 2048 bits are \
-                     not supported"
-                        .to_string(),
-                ));
-            }
-            total_size += reg_size;
-        }
-        if total_size != inner.data.len() {
-            Err(VersionizeError::Deserialize(
-                "Failed to deserialize aarch64 registers. Sum of registers sizes is not equal to \
-                 registers data length"
-                    .to_string(),
-            ))
-        } else {
-            Ok(Self { inner })
-        }
```
I had an offline discussion with @ShadowCurse about this. Apparently, the check for <= 2048 bits made sense only for cases where we would have snapshots from a different version than the one we are loading the snapshot for. But I realize now that the check for the total size of the registers might be important (?)

ShadowCurse replied:
Technically this check is to prevent snapshot tampering, but we already have crc check for it, so I think we can remove it. Also snapshots are generally trusted by us.

roypat replied:
I remember that we needed this check to get Firecracker to gracefully exit instead of exit with panic if an invalid snapshot file is passed. If that's no longer the case, then we can remove it, but if without this, Firecracker can be made to panic, we should keep it

bchalios replied:
Under which condition would this cause Firecracker to panic? Right now, we only snapshot and restore from the same Firecracker version. Under this condition, I don't see how this could fail. Am I missing something?

ShadowCurse replied:
Oh yeah, I think I remember now. It is because of this [`unreachable`](https://github.com/firecracker-microvm/firecracker/blob/bccf2d4ebaa54265584fb23898fdb43cf92bd75c/src/vmm/src/arch/aarch64/regs.rs#L171 ) If snapshot was modified and the id of some register is changed to have values bigger than 8 in the reg size position then conversion will fail. (8 because 2^8 = 256 bytes -> 2048 bit registers - the max size kernel supports)

bchalios replied:
Ok, I will just re-introduce the manual deserialization. It's not that much of a trouble and we will avoid any surprises in the future.

bchalios replied:
Ok, manual implementations are not as straight-forward as I thought. I will look into it again after I address the rest of the comments.

ShadowCurse replied:
Maybe you can leave the wrapper type for now and we create a `good new issue` to remove it?

bchalios replied:
Ok, done! I combined your idea Egor with an idea of Patrick's where I just serialize manually a tuple including the fields and then deserialize that tuple and perform the checks.

ShadowCurse commented on src/snapshot/src/lib.rs:
```diff
@@ -158,432 +160,215 @@ impl Snapshot {
+            }))
+        ));
+
+        // We can't support minor versions bigger than ours
+        let snapshot = Snapshot::new(Version::new(1, 2, 12));
```
Shouldn't this be `1.4.12` that is not supported because it is bigger than `1.3.12`?

bchalios replied:
`1.2.12` here represents the snapshot version that a Firecracker binary supports. `1.3.12` is the version of the saved snapshot.

ShadowCurse replied:
ok, I get it now.

ShadowCurse commented on src/snapshot/src/lib.rs:
```diff
@@ -158,432 +160,215 @@ impl Snapshot {
+            }))
+        ));
+
+        // But we can support minor versions smaller or equeal to ours. We also support
+        // all patch versions within our supported major.minor version.
+        let snapshot = Snapshot::new(Version::new(1, 4, 12));
```
And here it should be `1.2.12` as supported as it is less than `1.3.12`

bchalios replied:
See last comment

bchalios commented on Cargo.lock:
```diff
@@ -1413,7 +1421,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 dependencies = [
  "bincode",
- "crc64",
+ "crc64 1.0.0",
```
Yup, I think this was due to the fact that we kept using `Vesionize` which was bringing its own version of the crate. I didn't remove it at the time.

bchalios commented on src/vmm/src/arch/aarch64/regs.rs:
```diff
@@ -244,7 +243,7 @@ impl Aarch64RegisterVecInner {
+#[derive(Default, Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub struct Aarch64RegisterVec {
     inner: Aarch64RegisterVecInner,
 }
```
done

bchalios commented on tests/integration_tests/functional/test_cmd_line_parameters.py:
```diff
@@ -36,8 +36,10 @@ def test_describe_snapshot_all_versions(
 
-    # Fetch Firecracker binary for the latest version
-    fc_binary, _ = get_firecracker_binaries()
+    # Starting from version 1.6.0, each Firecracker binary supports a (potentially)
+    # different snapshot data format version. So, here we use the binary that created
+    # the snapshot to check the --describe-snapshot command.
+    fc_binary = firecracker_release.path
```
Problem is it can't. The snapshot header has changed in a way that it doesn't allow it parse older snapshot versions.

But you are right. I think the correct thing to do here is to skip this for older supported releases. WDYT?

roypat replied:
> I think the correct thing to do here is to skip this for older supported releases.

Yeah, this is what I meant, sorry :)

bchalios replied:
Done :)

roypat replied:
Shouldn't the diff to the test be?

```diff
diff --git a/tests/integration_tests/functional/test_cmd_line_parameters.py b/tests/integration_tests/functional/test_cmd_line_parameters.py
index 59e26e4e9..99d17ca03 100644
--- a/tests/integration_tests/functional/test_cmd_line_parameters.py
+++ b/tests/integration_tests/functional/test_cmd_line_parameters.py
@@ -7,26 +7,22 @@ from pathlib import Path
 
 import pytest
 
-from framework.utils import run_cmd
+from framework.utils import run_cmd, get_firecracker_version_from_toml
 from host_tools.cargo_build import get_firecracker_binaries
 from host_tools.metrics import validate_fc_metrics
 
 
-def test_describe_snapshot_all_versions(
-    microvm_factory, guest_kernel, rootfs, firecracker_release
-):
+def test_describe_snapshot_all_versions(microvm_factory, guest_kernel, rootfs):
     """
     Test `--describe-snapshot` correctness for all snapshot versions.
 
     For each release create a snapshot and verify the data version of the
     snapshot state file.
     """
-    target_version = firecracker_release.snapshot_version
+    target_version = get_firecracker_version_from_toml()
     vm = microvm_factory.build(
         guest_kernel,
         rootfs,
-        fc_binary_path=firecracker_release.path,
-        jailer_binary_path=firecracker_release.jailer,
     )
     vm.spawn()
     vm.basic_config(track_dirty_pages=True)

```

bchalios replied:
I _think_ not. IIUC the point of this test, it is to check if `firecracker --describe-snapshot` works correctly for all previously supported snapshot formats. Once we start having more Firecracker binaries that "speak" the new snapshot format this test will create a snapshot for every previously supported Firecracker version (>= 1.6.0) and then use the latest `main` to check that `--describe-snapshot` works for all of the previous snapshots.

roypat replied:
Ahhhh, I get it now. Didnt catch that it used the current firecracker version to call `--describe-snapshot`. Okay discard what I said :D

bchalios commented on tests/integration_tests/functional/test_snapshot_basic.py:
```diff
@@ -53,18 +48,20 @@ def test_snapshot_current_version(uvm_nano):
     fc_binary, _ = get_firecracker_binaries()
+    # Get supported snapshot version from Firecracker binary
+    snapshot_version = (
+        run_cmd([(str(fc_binary)), "--snapshot-version"]).stdout.strip().split("\n")[0]
```
It seems that `run_cmd_sync` does two different things depending on whether the command we pass is a list or not, although I'm not sure what's the difference between the two.

bchalios replied:
In any case, I did the change :P

bchalios commented on tests/integration_tests/performance/test_snapshot_perf.py:
true! Are there not unused code warnings for Python?

roypat replied:
Sadly no :(
I suppose its because you can technically do something absurd such as constructing a function name via string concatenation, then call it via looking it up in `__dict__` or `globals()`. So you can only find out which functions are called by actually running the code. 
That's why we had PRs in the past that deleted like 500 lines of unused python code :joy:

bchalios commented on tests/integration_tests/functional/test_snapshot_advanced.py:
true! Hadn't noticed. Removed.

bchalios commented on docs/snapshotting/snapshot-support.md:
```diff
@@ -174,41 +174,13 @@ The snapshot functionality is still in developer preview due to the following:
-version) will be defined later.
+The microVM state snapshot file uses a data format that follows semantic
+versioning. Each Firecracker binary supports a fixed version of the snapshot
+data format. When creating the snapshots it will use the supported data format
```
done!

bchalios commented on src/snapshot/src/lib.rs:
```diff
@@ -1,154 +1,156 @@
-    if magic_arch == BASE_MAGIC_ID {
-        return Ok((magic_id & !BASE_MAGIC_ID_MASK) as u16);
+impl SnapshotHdr {
+    fn new(version: &Version) -> Self {
+        Self {
+            magic: SNAPSHOT_MAGIC_ID,
+            version: version.clone(),
+        }
```
done

bchalios replied:
fixed this instance as well.

bchalios commented on tests/framework/artifacts.py:
```diff
@@ -99,7 +99,20 @@ def version_tuple(self):
+        # independent of Firecracker versions. For these Firecracker versions, use
+        # the --snapshot-version Firecracker flag, to figure out which snapshot version
+        # it supports.
+        if packaging.version.parse(self.version) < packaging.version.parse("1.6.0"):
+            return self.version_tuple[:2] + (0,)
```
I added a TODO in both places where I do these checks. I will open the Github issues to remove these, once 1.5 goes out of support.

roypat commented:
I think https://github.com/firecracker-microvm/firecracker/blob/da92bf6d4afa678565dcf9df2e4e00d2fa273d96/src/vmm/src/persist.rs#L162-L165 can also be deleted :)

bchalios commented on src/vmm/src/arch/aarch64/regs.rs:
```diff
@@ -241,85 +236,42 @@ impl Aarch64RegisterVecInner {
-                    "Failed to deserialize aarch64 registers. Registers bigger then 2048 bits are \
-                     not supported"
-                        .to_string(),
+            if reg_size >= RegSize::U2048_SIZE {
```
done!

bchalios commented on Cargo.toml:
```diff
@@ -11,4 +11,5 @@ panic = "abort"
 lto = true
```
done!

ShadowCurse commented on src/vmm/src/devices/virtio/persist.rs:
```diff
@@ -337,35 +319,24 @@ mod tests {
 
-        let restored_state =
-            VirtioDeviceState::deserialize(&mut mem.as_slice(), &version_map, 1).unwrap();
+        let restored_state: VirtioDeviceState = Snapshot::deserialize(&mut mem.as_slice()).unwrap();
```
Maybe create an issue to remove these later to limit changes in this PR?

roypat commented:
@ShadowCurse do you know if the "needed for compatibility" enum variants in https://github.com/firecracker-microvm/firecracker/blob/main/src/vmm/src/cpu_config/aarch64/static_cpu_templates/mod.rs#L14-L19 can be removed now that versionize is going away?

roypat commented on src/vmm/src/vmm_config/boot_source.rs:
```diff
@@ -32,7 +30,6 @@ pub struct BootSourceConfig {
     pub initrd_path: Option<String>,
     /// The boot arguments to pass to the kernel. If this field is uninitialized,
     /// DEFAULT_KERNEL_CMDLINE is used.
-    #[serde(skip_serializing_if = "Option::is_none")]
```
mh, why remove this? Or rather, why was this here in the first place?

Actually, why did these already implement Serialize anyway. They're for the API, so only Deserialize should've been needed :thinking:

roypat commented on src/vmm/src/vstate/memory.rs:
```diff
@@ -525,6 +523,57 @@ mod tests {
+        let page_size = get_page_size().unwrap();
+        let region_size = page_size * 3;
+
+        // Test with a single region
+        let guest_memory =
+            GuestMemoryMmap::from_raw_regions(&[(GuestAddress(0), region_size)], false).unwrap();
+        check_serde(&guest_memory);
+
+        // Test with some regions
+        let regions = vec![
+            (GuestAddress(0), region_size),                      // pages 0-2
+            (GuestAddress(region_size as u64), region_size),     // pages 3-5
+            (GuestAddress(region_size as u64 * 2), region_size), // pages 6-8
+        ];
+        let guest_memory = GuestMemoryMmap::from_raw_regions(&regions, true).unwrap();
+        check_serde(&guest_memory);
+
+        // Test with dirty memory
+        let dirty_map = [
+            // page 0: not dirty
+            (0, page_size, false),
+            // pages 1-2: dirty range in one region
+            (page_size, page_size * 2, true),
+            // page 3: not dirty
+            (page_size * 3, page_size, false),
+            // pages 4-7: dirty range across 2 regions,
+            (page_size * 4, page_size * 4, true),
+            // page 8: not dirty
+            (page_size * 8, page_size, false),
+        ];
+
+        // Mark dirty memory
+        for (addr, len, dirty) in &dirty_map {
+            if *dirty {
+                guest_memory.mark_dirty(GuestAddress(*addr as u64), *len);
+            }
+        }
+
+        check_serde(&guest_memory);
+    }
```
Mh, do we even want to serialize the dirty bitmap? After a snapshot is taken it should get reset to "everything is clean" (because we just took a snapshot)

bchalios replied:
You are right. We don't want to, and in fact we don't do it. This check does absolutely nothing :) I will remove

roypat commented:
You missed a handful of the "requires version bump" comments, I think:

```
$ grep "version bump" src -R
src/vmm/src/arch/aarch64/gic/gicv2/regs/dist_regs.rs:// are not saved. NOTICE: Any changes to this structure require a snapshot version bump.
src/vmm/src/arch/aarch64/gic/gicv2/regs/icc_regs.rs:// NOTICE: Any changes to this structure require a snapshot version bump.
src/vmm/src/arch/aarch64/gic/gicv3/regs/redist_regs.rs:// NOTICE: Any changes to this structure require a snapshot version bump.
src/vmm/src/arch/aarch64/gic/gicv3/regs/redist_regs.rs:// NOTICE: Any changes to this structure require a snapshot version bump.
src/vmm/src/arch/aarch64/gic/gicv3/regs/dist_regs.rs:// snapshot version bump.
src/vmm/src/arch/aarch64/gic/gicv3/regs/icc_regs.rs:// NOTICE: Any changes to this structure require a snapshot version bump.
src/vmm/src/arch/aarch64/gic/gicv3/regs/icc_regs.rs:// NOTICE: Any changes to this structure require a snapshot version bump.
```

roypat commented on src/vmm/src/device_manager/persist.rs:
```diff
@@ -173,9 +166,9 @@ pub struct ConnectedLegacyState {
-#[derive(Debug, Clone, PartialEq, Eq, Versionize)]
+#[derive(Debug, Default, Clone, PartialEq, Eq, Serialize, Deserialize)]
 pub enum MmdsVersionState {
+    #[default]
```
maybe `V2` is a more sensible default?

bchalios replied:
In fact, we should completely remove the `Default` implementation here. The state will always be stored in the snapshot file. It doesn't make sense to have this at all, since we will never trigger it.

roypat commented on tests/integration_tests/functional/test_api.py:
```diff
@@ -1193,6 +1193,7 @@ def test_get_full_config_after_restoring_snapshot(microvm_factory, uvm_nano):
     expected_cfg["boot-source"] = {
         "kernel_image_path": uvm_nano.get_jailed_resource(uvm_nano.kernel_file),
         "initrd_path": None,
+        "boot_args": None,
```
ah, this answers my first question :)

I guess this is fine because we are adding a new field to an existing response, which can be done in a minor

ShadowCurse commented:
> @ShadowCurse do you know if the "needed for compatibility" enum variants in https://github.com/firecracker-microvm/firecracker/blob/main/src/vmm/src/cpu_config/aarch64/static_cpu_templates/mod.rs#L14-L19 can be removed now that versionize is going away?

Hmm... actually by removing `verisonize` we break compatibility, so I think this is a good time to fix `StaticCpuTemplate` enum (for both x86 and aarch64). The reason the enum looks so strange is the historical reason that goes back to the time when we did not have aarch64 templates, but still were storing `StaticCpuTemplate` in the aarch64 snapshots.

I think we should clean `StaticCpuTemplate` from `Empty#` fields and `None` case maybe. But this can be done in a separate PR (it is not really a `versionize` problem). We just need not to forget.

roypat commented:
> > @ShadowCurse do you know if the "needed for compatibility" enum variants in https://github.com/firecracker-microvm/firecracker/blob/main/src/vmm/src/cpu_config/aarch64/static_cpu_templates/mod.rs#L14-L19 can be removed now that versionize is going away?
> 
> Hmm... actually by removing `verisonize` we break compatibility, so I think this is a good time to fix `StaticCpuTemplate` enum (for both x86 and aarch64). The reason the enum looks so strange is the historical reason that goes back to the time when we did not have aarch64 templates, but still were storing `StaticCpuTemplate` in the aarch64 snapshots.
> 
> I think we should clean `StaticCpuTemplate` from `Empty#` fields and `None` case maybe. But this can be done in a separate PR (it is not really a `versionize` problem). We just need not to forget.

I actually already have a local branch that does both, I just need to publish it once this PR is merged xP

bchalios commented:
Changes from v1 -> v2:

* Removed comments I had missed about bumping Versionize version if structs changed
* Removed unnecessary check for serialization/deserialization of GuestMemory dirty tracking bitmap
* Removed default implementation for `MmdsVersionState`. We will always get a valid value from the snapshot so the default implementation was dead code

roypat commented on src/vmm/src/persist.rs:
```diff
@@ -48,39 +47,18 @@ use crate::{mem_size_mib, vstate, EventManager, Vmm, VmmError};
 const FC_V0_23_MAX_DEVICES: u32 = 11;
```
I think this constant can go away :thinking:

bchalios replied:
yup!

bchalios commented:
Changes v3 -> v4:

* Removed `FC_V0_23_MAX_DEVICES` which was not used any more and an associated Error variant that was only there to express that we were asked to create more than `FC_V0_23_MAX_DEVICES` devices.

pb8o commented on tests/integration_tests/functional/test_snapshot_basic.py:
```diff
@@ -53,18 +48,20 @@ def test_snapshot_current_version(uvm_nano):
 
     # Fetch Firecracker binary for the latest version
     fc_binary = uvm_nano.fc_binary_path
+    # Get supported snapshot version from Firecracker binary
+    snapshot_version = (
+        run_cmd(f"{fc_binary} --snapshot-version").stdout.strip().splitlines()[0]
+    )
```
nit: this could be a method of Microvm class. But since it's not called often it may not be worth it.

pb8o commented on src/snapshot/benches/version_map.rs:
```diff
@@ -1,133 +0,0 @@
-// Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Oh removing this will need a change in https://github.com/firecracker-microvm/firecracker/blob/main/tools/bump-version.sh to stop looking for versions in this file. I think it's the only one.

bchalios replied:
Very nice catch. Can you think of another check to replace this with, or should we just drop the check all together?

bchalios replied:
Also, is there a way to test the release scripts?

roypat replied:
We have a "release sanity build" step in our CI, but it only triggers if specific files in `tools` are touched. You could add a dummy commit that touches `tools/devtool` to trigger it

pb8o replied:
I would go with dropping the check altogether. The issue that this was checking for is gone with your PR, so we don't need it anymore. You can test the release scripts locally by `devtool build --release`, but if you drop the check it will get done in the PR anyway.

bchalios replied:
Ok, done

bchalios replied:
it might be worth to run the check in every PR (?)

bchalios commented on src/snapshot/src/crc.rs:
```diff
@@ -0,0 +1,159 @@
+// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
What year do we typically put in there? Is it the first time this was written, or the last time it was touched?

bchalios replied:
Answering my own question: this should be the year that we first published this code. This was taken from the versionize crate: https://github.com/firecracker-microvm/versionize/blob/main/src/crc.rs, so I think that I should just leave 2020. WDYT?

pb8o replied:
Ah that's fine by me... I think I have also seen 2020-2024, but it would be the only one in the repo to have a year range.

bchalios replied:
Ok, I stayed with the 2020 year. Another issue going the year range is that it would imply that we need to update the range in all the files once per year.

bchalios commented:
Changes v4 -> v5:

* Dropped check for versions existing in `src/vmm/src/version_map.rs`; the file doesn't exist any more.
* Added correct year (2020) in copyright notice of `src/vmm/src/snapshot/src/crc.rs` file. This file was taken from the `versionize` crate implementation and 2020 was the first year of publication.
